rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -606,6 +607,7 @@ class ClientListener : public SubscriberListener
           } else {
             list.push_back(response);
           }
+          list_has_data_.store(true);
```

Comment:
With the `list_has_data_` flag now being changed outside of the `conditionMutex_` lock this introduces a race condition with the `rmw_wait()` function. After the function locked the mutex it checks `hasData()` of the listeners and then makes a decision if it has to call `wait`. With this change the boolean flag might change at any time which make `rmw_wait()` call `wait` whe it shouldn't.

See ros2/rmw_fastrtps#147 for the proposed fix.

Before this patch the flag was represented by the state of the container and therefore only changed within the scope of the locked condition.